File: BrazeProject/ios/BrazeReactGIFHelper.m
#import "BrazeReactGIFHelper.h"
#import <SDWebImage/SDAnimatedImage.h>
#import <SDWebImage/SDAnimatedImageView.h>

@import BrazeUI;

// Braze InAppMessage GIF Support
@implementation BrazeReactGIFHelper

+ (void)setGIFProvider {
  BRZGIFViewProvider.shared = [BrazeReactGIFHelper sdWebImage];
}

+ (BRZGIFViewProvider *)sdWebImage {
  return [BRZGIFViewProvider
          providerWithView:^UIView *_Nonnull(NSURL *_Nullable url) {
    UIImage *image = [BrazeReactGIFHelper imageForURL:url];
    SDAnimatedImageView *view =
    [[SDAnimatedImageView alloc] initWithImage:image];
    return view;
  }
          updateView:^(UIView *_Nonnull view, NSURL *_Nullable url) {
    if ([view isKindOfClass:[SDAnimatedImageView class]]) {
      ((SDAnimatedImageView *)(view)).image =
      [BrazeReactGIFHelper imageForURL:url];
    }
  }];
}

+ (UIImage *)imageForURL:(NSURL *)url {
  if (url == nil) {
    return nil;
  }
  if ([url.pathExtension isEqualToString:@"gif"]) {
    return [SDAnimatedImage imageWithContentsOfFile:url.path];
  } else {
    return [UIImage imageWithContentsOfFile:url.path];
  }
}

@end

File: BrazeProject/ios/BrazeReactGIFHelper.swift
import UIKit
import BrazeUI
import SDWebImage

// Braze In-App Message GIF Support
extension GIFViewProvider {
  
  /// A GIF view provider using [SDWebImage](https://github.com/SDWebImage/SDWebImage) as a
  /// rendering library.
  public static let sdWebImage = Self(
    view: { SDAnimatedImageView(image: image(for: $0)) },
    updateView: { ($0 as? SDAnimatedImageView)?.image = image(for: $1) }
  )
  
  private static func image(for url: URL?) -> UIImage? {
    guard let url else { return nil }
    return url.pathExtension == "gif"
      ? SDAnimatedImage(contentsOfFile: url.path)
      : UIImage(contentsOfFile: url.path)
  }
  
}
